src/v/model/record_batch_reader.h

```diff
@@ -280,6 +280,36 @@ make_memory_record_batch_reader(model::record_batch b) {
     return make_memory_record_batch_reader(std::move(batches));
 }
 
+// clang-format off
+template<typename Func>
+CONCEPT(requires requires(Func f, model::record_batch&& batch) {
+    { f(batch) } -> std::same_as<model::record_batch>;
+})
+// clang-format on
+ss::future<record_batch_reader::data_t> transform_reader_to_memory(
+  record_batch_reader reader, timeout_clock::time_point timeout, Func&& f) {
+    using data_t = record_batch_reader::data_t;
+    class consumer {
+    public:
+        explicit consumer(Func&& f)
+          : _func(std::forward<Func>(f)) {}
```

Comment:
`_func` is stored by value, and you don't have a forwarding reference here, so this should be `std::move`.